foreman/api/discovery_rule.go

```diff
@@ -0,0 +1,211 @@
+package api
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/HanseMerkur/terraform-provider-utils/log"
+)
+
+const (
+	DiscoveryRuleEndpointPrefix = "v2/discovery_rules"
+)
+
+type ForemanDiscoveryRule struct {
+	// Inherits the base object's attributes
+	ForemanObject
+	Name                  string `json:"name"`
+	Search                string `json:"search,omitempty"`
+	HostGroupId           int    `json:"hostgroup_id,omitempty"`
+	Hostname              string `json:"hostname,omitempty"`
+	HostLimitMaxCount     int    `json:"max_count,omitempty"`
+	Priority              int    `json:"priority"`
+	Enabled               bool   `json:"enabled"`
+	LocationIds           []int  `json:"location_ids,omitempty"`
+	OrganizationIds       []int  `json:"organization_ids,omitempty"`
+	DefaultLocationId     int    `json:"location_id,omitempty"`
+	DefaultOrganizationId int    `json:"organization_id,omitempty"`
+}
+
+// -----------------------------------------------------------------------------
+// CRUD Implementation
+// -----------------------------------------------------------------------------
+
+func (c *Client) CreateDiscoveryRule(ctx context.Context, d *ForemanDiscoveryRule) (*ForemanDiscoveryRule, error) {
+	log.Tracef("foreman/api/discovery_rule.go#Create")
+
+	reqEndpoint := fmt.Sprintf("/%s", DiscoveryRuleEndpointPrefix)
+
+	if d.DefaultLocationId == 0 {
+		d.DefaultLocationId = c.clientConfig.LocationID
+	}
+
+	if d.DefaultOrganizationId == 0 {
+		d.DefaultOrganizationId = c.clientConfig.OrganizationID
+	}
+
+	dJSONBytes, jsonEncErr := c.WrapJSON("discovery_rule", d)
+	if jsonEncErr != nil {
+		return nil, jsonEncErr
+	}
+
+	log.Debugf("discoveryruleJSONBytes: [%s]", dJSONBytes)
+
+	req, reqErr := c.NewRequestWithContext(
+		ctx,
+		http.MethodPost,
+		reqEndpoint,
+		bytes.NewBuffer(dJSONBytes),
+	)
+	if reqErr != nil {
+		return nil, reqErr
+	}
+
+	var createdDiscoveryRule ForemanDiscoveryRule
+	sendErr := c.SendAndParse(req, &createdDiscoveryRule)
```

Comment:
```suggestion
	if err := c.SendAndParse(req, &createdDiscoveryRule); err != nil {
	  return nil, err
	}
```

Reply:
Also, it can be helpful to add some context to these errors: https://pkg.go.dev/fmt#Errorf